src/core/utils/debug_uart.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ *  Copyright (c) 2017, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <openthread/config.h>
+#include <stdio.h>
+#include <stdarg.h>
+
+#include <openthread/platform/toolchain.h>
+#include <openthread/platform/alarm-milli.h>
+#include <common/logging.hpp>
+#include <utils/debug_uart.h>
+
+/*
+ * Implimentation note:
+ *   These are all "weak" so that
+ *   a platform may if it chooses
+ *   override the instance.
+ */
+
+OT_TOOL_WEAK
+void otPlatDebugUart_printf(const char *fmt, ...)
```

Comment:
Here, and other functions in this (and the matching header file) do not exactly follow the openthread naming style.

On one hand, these function generally act 100% as their **VERY WELL KNOWN** (named) public counter parts act however, so one thinking is we make those function names part of this function name.

Other names could be:

* otPlatDebugUartPrintf() [better, includes the name "printf"]
* or  otPlatDebugUartFormat() [not so great, as a human I'd look for that very well known Printf function name]

On the other hand, that does not fit the Openthread Model.

We thought it best to have the discussion about these function names as part of the pull request process.